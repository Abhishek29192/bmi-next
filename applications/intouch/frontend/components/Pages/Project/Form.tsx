import React from "react";
import { useTranslation } from "next-i18next";
import { Guarantee } from "@bmi/intouch-api-types";
import { Form, FormProps } from "@bmi/components";
import { TextField } from "@bmi/components";
import { Grid } from "@bmi/components";
import { Typography } from "@bmi/components";
import { getDateOnlyString } from "../../../lib/utils/date";
import { findProjectGuarantee } from "../../../lib/utils/project";
import { GetProjectQuery } from "../../../graphql/generated/operations";
import { getNestedValue } from "../../../lib/utils/object";
import TechnologyInput from "./TechnologyInput";
// TODO: move/split styles?
import styles from "./CreateProject/styles.module.scss";

export const getFieldLabel = (t, fieldName: string) =>
  t(`project-page:addProject.dialog.form.fields.${fieldName}`);
export const getFieldValueLabel = (t, fieldName: string, value: string) =>
  t(`project-page:addProject.dialog.form.values.${fieldName}.${value}`);

export const getFieldProps = (t, fieldName: string) => ({
  className: styles.input,
  name: fieldName,
  label: getFieldLabel(t, fieldName),
  fullWidth: true,
  fieldIsRequiredError: t("common:error_messages.required")
});

export const getValue = (project, fieldName) => {
  if (["startDate", "endDate"].includes(fieldName)) {
    return getDateOnlyString(getNestedValue(project, fieldName));
  }

  return getNestedValue(project, fieldName);
};

// If there is a guarantee present, cannot edit these fields
const disabledFields = [
  "name",
  "description",
  "roofArea",
  "startDate",
  "endDate",
  "siteAddress.firstLine",
  "siteAddress.secondLine",
  "siteAddress.town",
  "siteAddress.postcode",
  "siteAddress.region",
  "siteAddress.country"
];

export const isFieldDisabled = (
  guarantee: Pick<Guarantee, "status">,
  fieldName: string
) => {
  if (!guarantee) {
    return false;
  }

  // Cannot change technology as soon as there is a guarantee in progress
  if (guarantee && fieldName === "technology") {
    return true;
  }

  if (["APPROVED", "SUBMITTED", "REVIEW"].includes(guarantee.status)) {
    return disabledFields.includes(fieldName);
  }

  return false;
};

type ProjectFormProps = {
  project?: GetProjectQuery["project"];
  onSubmit?: FormProps["onSubmit"];
  isSubmitting?: boolean;
};

const ProjectForm = ({ project, onSubmit, isSubmitting }: ProjectFormProps) => {
  const { t } = useTranslation("project-page");

  const guarantee = findProjectGuarantee(project);

  const getFieldPropsFn = project
    ? (t, fieldName) => {
        return {
          ...getFieldProps(t, fieldName),
          defaultValue: getValue(project, fieldName),
          disabled: isFieldDisabled(guarantee, fieldName)
        };
      }
    : getFieldProps;

  return (
    <Form className={styles.form} onSubmit={onSubmit} rightAlignButton>
      <Typography variant="h5" className={styles.sectionTitle}>
        {t("project-page:addProject.dialog.sections.projectDetails.title")}
      </Typography>

      <TextField {...getFieldPropsFn(t, "name")} isRequired />

      <div className={styles.input}>
        <Typography variant="h6" className={styles.sectionTitle}>
          {getFieldLabel(t, "technology")}
        </Typography>
        <TechnologyInput {...getFieldPropsFn(t, "technology")} isRequired />
      </div>

      <TextField
        {...getFieldPropsFn(t, "roofArea")}
        isRequired
        type="number"
        inputProps={{
          min: 1,
          step: 1
        }}
      />

      <TextField {...getFieldPropsFn(t, "description")} isTextArea rows={6} />

      <Typography variant="h6" className={styles.sectionTitle}>
        {t("project-page:addProject.dialog.sections.dates.title")}
      </Typography>
      <Grid container>
        <Grid item sm={6}>
          <TextField
            {...getFieldPropsFn(t, "startDate")}
            type="date"
            InputLabelProps={{
              shrink: true
            }}
            isRequired
          />
          <TextField
            {...getFieldPropsFn(t, "endDate")}
            type="date"
            InputLabelProps={{
              shrink: true
            }}
            isRequired
          />
        </Grid>
      </Grid>

      <Typography variant="h5" className={styles.sectionTitle}>
        {t("project-page:addProject.dialog.sections.siteAddress.title")}
      </Typography>
      <TextField {...getFieldPropsFn(t, "siteAddress.firstLine")} isRequired />
      <TextField {...getFieldPropsFn(t, "siteAddress.secondLine")} />
      <TextField {...getFieldPropsFn(t, "siteAddress.town")} isRequired />
      <Grid container>
        <Grid item lg={6}>
          <TextField
            {...getFieldPropsFn(t, "siteAddress.postcode")}
            isRequired
          />
        </Grid>
      </Grid>
      <TextField {...getFieldPropsFn(t, "siteAddress.region")} />
      <TextField {...getFieldPropsFn(t, "siteAddress.country")} isRequired />

      <Form.ButtonWrapper>
        <Form.SubmitButton disabled={isSubmitting}>
          {t("project-page:addProject.dialog.form.actions.submit")}
        </Form.SubmitButton>
      </Form.ButtonWrapper>
    </Form>
  );
};

export default ProjectForm;
