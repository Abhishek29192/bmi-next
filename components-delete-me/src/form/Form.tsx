import classnames from "classnames";
import React, { FormEvent, useState } from "react";
import { useStyles } from "./styles";
import { InputValue } from "./withFormControl";
import Button from "./_Button";
import SubmitButton from "./_SubmitButton";

export type Values = Record<string, InputValue | undefined>;
type ValidationPasses = false | "" | null | undefined;
export type ValidationResult = string | ValidationPasses;
type Errors = Record<string, ValidationResult>;

export type Props = Omit<React.HTMLProps<HTMLFormElement>, "onSubmit"> & {
  children: React.ReactNode;
  onSubmit?: (event: FormEvent<HTMLFormElement>, values: Values) => void;
  submitButtonLabel?: string;
  rightAlignButton?: boolean;
  buttonClassName?: string;
};

type ContextType = {
  updateFormState: (fieldValues: Values, fieldErrors?: Errors) => void;
  hasBeenSubmitted: boolean;
  submitButtonDisabled: boolean;
  values: Values;
};

export const FormContext = React.createContext<ContextType>({
  updateFormState: () => {
    // no-op
  },
  hasBeenSubmitted: false,
  submitButtonDisabled: false,
  values: {}
});

const Form = ({
  children,
  onSubmit,
  rightAlignButton,
  buttonClassName,
  className,
  ...formProps
}: Props) => {
  const classes = useStyles();
  const [values, setValues] = useState<Values>({});
  const [errors, setErrors] = useState<Errors>({});
  const [hasBeenSubmitted, setHasBeenSubmitted] = useState(false);
  const submitButtonDisabled = Object.values(errors).some(Boolean);

  const updateFormState = (fieldValues: Values, fieldErrors: Errors = {}) => {
    setValues((prev) => ({ ...prev, ...fieldValues }));
    setErrors((prev) => ({ ...prev, ...fieldErrors }));
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    setHasBeenSubmitted(true);
    if (onSubmit) {
      onSubmit(event, values);
    }
  };

  return (
    <FormContext.Provider
      value={{
        updateFormState,
        hasBeenSubmitted,
        submitButtonDisabled,
        values
      }}
    >
      <form
        onSubmit={handleSubmit}
        className={classnames(
          classes.root,
          className,
          rightAlignButton && classes.rightAlignButton
        )}
        {...formProps}
      >
        {children}
      </form>
    </FormContext.Provider>
  );
};

const FormRow = ({ children }: { children: React.ReactNode }) => {
  const classes = useStyles();
  return <div className={classes.row}>{children}</div>;
};

const ButtonWrapper = ({ children }: { children: React.ReactNode }) => {
  const classes = useStyles();
  return <div className={classes.buttonWrapper}>{children}</div>;
};

Form.Row = FormRow;
Form.ButtonWrapper = ButtonWrapper;
Form.Button = Button;
Form.SubmitButton = SubmitButton;

export default Form;
