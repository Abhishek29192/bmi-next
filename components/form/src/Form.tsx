import classnames from "classnames";
import React, { FormEvent, useState } from "react";
import Button from "./_Button";
import styles from "./Form.module.scss";
import SubmitButton from "./_SubmitButton";
import { InputValue } from "./withFormControl";

type Values = Record<string, InputValue | undefined>;
type Errors = Record<string, string>;

export type Props = Omit<React.HTMLProps<HTMLFormElement>, "onSubmit"> & {
  children: React.ReactNode;
  onSubmit?: (event: FormEvent<HTMLFormElement>, values: Values) => void;
  submitButtonLabel?: string;
  rightAlignButton?: boolean;
  buttonClassName?: string;
};

type ContextType = {
  updateFormState: (fieldValues: Values, fieldErrors: Errors) => void;
  hasBeenSubmitted: boolean;
  submitButtonDisabled: boolean;
  values: object;
};

export const FormContext = React.createContext<ContextType>({
  updateFormState: (fieldValues, fieldErrors) => {},
  hasBeenSubmitted: false,
  submitButtonDisabled: false,
  values: {}
});

const Form = ({
  children,
  onSubmit,
  rightAlignButton,
  buttonClassName,
  className,
  ...formProps
}: Props) => {
  const [values, setValues] = useState<Values>({});
  const [errors, setErrors] = useState<Errors>({});
  const [hasBeenSubmitted, setHasBeenSubmitted] = useState(false);
  const submitButtonDisabled = Object.values(errors).some(Boolean);

  const updateFormState = (fieldValues: Values, fieldErrors: Errors) => {
    setValues((prev) => ({ ...prev, ...fieldValues }));
    setErrors((prev) => ({ ...prev, ...fieldErrors }));
  };

  const handleSubmit = (event: React.FormEvent<HTMLFormElement>) => {
    setHasBeenSubmitted(true);
    if (onSubmit) {
      onSubmit(event, values);
    }
  };

  return (
    <FormContext.Provider
      value={{
        updateFormState,
        hasBeenSubmitted,
        submitButtonDisabled,
        values
      }}
    >
      <form
        onSubmit={handleSubmit}
        className={classnames(styles["Form"], className, {
          [styles["Form--rightAlignButton"]!]: rightAlignButton
        })}
        {...formProps}
      >
        {children}
      </form>
    </FormContext.Provider>
  );
};

const FormRow = ({ children }: { children: React.ReactNode }) => {
  return <div className={styles["Row"]}>{children}</div>;
};

const ButtonWrapper = ({ children }: { children: React.ReactNode }) => {
  return <div className={styles["ButtonWrapper"]}>{children}</div>;
};

Form.Row = FormRow;
Form.ButtonWrapper = ButtonWrapper;
Form.Button = Button;
Form.SubmitButton = SubmitButton;

export default Form;
