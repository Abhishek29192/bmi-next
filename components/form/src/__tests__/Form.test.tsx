import "@testing-library/jest-dom";
import React, { useContext, useEffect } from "react";
import { fireEvent, render } from "@testing-library/react";
import Form, { FormContext, withFormControl } from "../";

const Component = () => {
  const { values, updateFormState } = useContext(FormContext);
  useEffect(() => {
    updateFormState(
      {
        test: "test"
      },
      {}
    );
  }, []);

  return <input data-testid="id" value={values["test"] as string} />;
};

describe("Form component", () => {
  it("renders correctly", () => {
    const { container } = render(
      <Form>
        <Form.Row>
          <p>Lorem Ipsum</p>
        </Form.Row>
        <Form.SubmitButton>Submit</Form.SubmitButton>
      </Form>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders with a multiple buttons", () => {
    const { container } = render(
      <Form>
        <Form.Row>
          <p>Lorem Ipsum</p>
        </Form.Row>
        <Form.ButtonWrapper>
          <Form.Button>Cancel</Form.Button>
          <Form.SubmitButton>Submit</Form.SubmitButton>
        </Form.ButtonWrapper>
      </Form>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders with a right aligned button", () => {
    const { container } = render(
      <Form rightAlignButton>
        <Form.Row>
          <p>Lorem Ipsum</p>
        </Form.Row>
        <Form.ButtonWrapper>
          <Form.SubmitButton>Submit</Form.SubmitButton>
        </Form.ButtonWrapper>
      </Form>
    );
    expect(container).toMatchSnapshot();
  });

  it("test submit form", () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(
      <Form onSubmit={onSubmit} data-testid="form">
        <Form.Row>
          <input type="text" name="text" data-testid="id" />
        </Form.Row>
        <Form.ButtonWrapper>
          <Form.SubmitButton>Submit</Form.SubmitButton>
        </Form.ButtonWrapper>
      </Form>
    );
    const textInput = getByTestId("id");
    fireEvent.change(textInput, {
      target: { value: "test" }
    });
    fireEvent.submit(getByTestId("form"));
    expect(onSubmit).toHaveBeenCalled();
  });

  it("test values passed to updateFormState function from context to be called on submit", () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(
      <Form onSubmit={onSubmit} data-testid="form">
        <Component />
      </Form>
    );
    const input = getByTestId("id");
    const form = getByTestId("form");
    fireEvent.click(input);
    fireEvent.submit(form);
    expect(onSubmit).toHaveBeenCalledWith(expect.any(Object), { test: "test" });
  });

  it("should not call onSubmit when it not pass for wrapped with formContext coponent", () => {
    const onSubmit = jest.fn();
    const { getByTestId } = render(
      <Form data-testid="form">
        <Component />
      </Form>
    );
    const form = getByTestId("form");
    fireEvent.submit(form);
    expect(onSubmit).not.toHaveBeenCalled();
  });

  it("test withFormControl calls onChange", () => {
    const mockComponent = jest.fn(({ ...props }) => (
      <input type="text" name="test" data-testid="id" {...props} />
    ));
    const WithFormControlComponent = withFormControl(mockComponent);
    const onChange = jest.fn();
    const { getByTestId, container } = render(
      <WithFormControlComponent
        name="test"
        onChange={onChange}
        defaultValue="Value"
      />
    );
    expect(container).toMatchSnapshot();
    const input = getByTestId("id");
    fireEvent.change(input, { target: { value: "Text" } });
    expect(onChange).toHaveBeenCalled();
  });

  it("test withFormControl not calls onChange", () => {
    const mockComponent = jest.fn(({ ...props }) => (
      <input type="text" name="test" data-testid="id" {...props} />
    ));
    const WithFormControlComponent = withFormControl(mockComponent);
    const onChange = jest.fn();
    const { getByTestId, container } = render(
      <WithFormControlComponent name="test" defaultValue="Value" />
    );
    expect(container).toMatchSnapshot();
    const input = getByTestId("id");
    fireEvent.change(input, { target: { value: "Text" } });
    expect(onChange).not.toHaveBeenCalled();
  });

  it("test withFormControl calls onBlur", () => {
    const mockComponent = jest.fn(({ ...props }) => (
      <input type="text" name="test" data-testid="id" {...props} />
    ));
    const getValidationErrorMock = jest.fn().mockImplementation(() => "error");
    const WithFormControlComponent = withFormControl(mockComponent);
    const { getByTestId } = render(
      <WithFormControlComponent
        name="test"
        isRequired={true}
        getValidationError={getValidationErrorMock}
        value="test"
      />
    );
    const input = getByTestId("id");
    fireEvent.blur(input);
    expect(input).toHaveAttribute("errortext", "error");
  });

  it("test withFormControl with fieldIsRequiredError", () => {
    const mockComponent = jest.fn(({ ...props }) => (
      <input type="text" name="test" data-testid="id" {...props} />
    ));
    const WithFormControlComponent = withFormControl(mockComponent);
    const { getByTestId } = render(
      <WithFormControlComponent
        name="test"
        fieldIsRequiredError="fieldIsRequiredError"
        isRequired={true}
      />
    );
    const input = getByTestId("id");
    expect(input).toHaveAttribute("errortext", "fieldIsRequiredError");
  });

  it("test withFormControl with no fieldIsRequiredError and value", () => {
    const mockComponent = jest.fn(({ ...props }) => (
      <input type="text" name="test" data-testid="id" {...props} />
    ));
    const WithFormControlComponent = withFormControl(mockComponent);
    const { container } = render(
      <WithFormControlComponent name="test" isRequired={true} />
    );
    expect(container).toMatchSnapshot();
  });
});
