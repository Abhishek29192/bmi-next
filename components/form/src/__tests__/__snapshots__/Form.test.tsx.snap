// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form component renders correctly 1`] = `
<form
  class="Form"
>
  <div
    class="Row"
  >
    <p>
      Lorem Ipsum
    </p>
  </div>
  <button
    class="MuiButtonBase-root MuiButton-root buttonStyles-root-1 MuiButton-contained buttonStyles-contained-3 Button MuiButton-containedPrimary"
    tabindex="0"
    type="submit"
  >
    <span
      class="MuiButton-label buttonStyles-label-4"
    >
      Submit
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</form>
`;

exports[`Form component renders with a multiple buttons 1`] = `
<form
  class="Form"
>
  <div
    class="Row"
  >
    <p>
      Lorem Ipsum
    </p>
  </div>
  <div
    class="ButtonWrapper"
  >
    <button
      class="MuiButtonBase-root MuiButton-root buttonStyles-root-20 MuiButton-contained buttonStyles-contained-22 Button MuiButton-containedPrimary"
      tabindex="0"
      type="button"
    >
      <span
        class="MuiButton-label buttonStyles-label-23"
      >
        Cancel
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
    <button
      class="MuiButtonBase-root MuiButton-root buttonStyles-root-20 MuiButton-contained buttonStyles-contained-22 Button MuiButton-containedPrimary"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label buttonStyles-label-23"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</form>
`;

exports[`Form component renders with a right aligned button 1`] = `
<form
  class="Form Form--rightAlignButton"
>
  <div
    class="Row"
  >
    <p>
      Lorem Ipsum
    </p>
  </div>
  <div
    class="ButtonWrapper"
  >
    <button
      class="MuiButtonBase-root MuiButton-root buttonStyles-root-39 MuiButton-contained buttonStyles-contained-41 Button MuiButton-containedPrimary"
      tabindex="0"
      type="submit"
    >
      <span
        class="MuiButton-label buttonStyles-label-42"
      >
        Submit
      </span>
      <span
        class="MuiTouchRipple-root"
      />
    </button>
  </div>
</form>
`;
